darkdetect/_base_listener.py

```diff
@@ -0,0 +1,118 @@
+    An abstract listener class
+    It is safe to call stop from a different thread than listen() was called in
+    provided multiple threads are not racing to call these methods
+    """
+
+    def __init__(self, callback: Optional[Callable[[str], None]]):
+        """
+        :param callback: The callback to use when the listener detects something
+        """
+        self._state: ListenerState = ListenerState.Dead
+        self.callback: Optional[Callable[[str], None]] = callback
+
+    def listen(self):
+        """
+        Start the listener if it is not already running
+        """
+        if self._state == ListenerState.Listening:
+            raise RuntimeError("Do not run .listen() from multiple threads concurrently")
+        if self._state == ListenerState.Stopping:
+            raise RuntimeError("Call .stop to wait for the previous listener to finish shutting down")
+        self._state = ListenerState.Listening
+        try:
+            self._listen()
+        except BaseException as e:
+            self._on_listen_fail_base(e)
+
+    def stop(self, timeout: Optional[int]) -> bool:
+        """
+        Initiate the listener stop sequence, wait at most timeout seconds for it to complete.
+        After this function returns, new theme changes will not invoke callbacks.
+        Running callbacks will not be interrupted.
+        May safely be called as many times as desired.
+        :param timeout: How many seconds to wait until the listener stops; None means infinite
+        :return: True if the listener completes before the timeout expires, else False
+        """
+        if timeout is not None and timeout < 0:
+            raise ValueError("timeout may not be negative")
+        if self._state == ListenerState.Listening:
+            self._initiate_shutdown()
+            self._state = ListenerState.Stopping
+        if self._state == ListenerState.Stopping:
+            if self._wait_for_shutdown(timeout):
+                self._state = ListenerState.Dead
+        return self._state == ListenerState.Dead
+
+    # Non-public helper methods
+
+    def _invoke_callback(self, value: str) -> None:
+        """
+        Invoke the stored callback if the state is listening
+        """
+        if self._state == ListenerState.Listening:
+            c: Optional[Callable[[str], None]] = self.callback
+            if c is not None:
+                c(value)
+
+    def _on_listen_fail_base(self, why: BaseException) -> None:
+        """
+        Invoked by .listen on all failures; self._state is unknown
+        Note that .stop may still be called on a failed listener!
+        Users should only override this if catching base exceptions is required
+        For example: Emergency cleanup when a user KeyboardInterrupts a program via Ctrl-C
+        :param why: The exception caught by _listen
+        """
+        if isinstance(why, (BaseException, NotImplementedError)):
+            raise why
+        self._on_listen_fail(why)
```

Comment:
What is the point of this `_on_listen_fail_base` vs `_on_listen_fail` separation? I'd rather unify the two by putting the code of the `base` method in the parent `_on_listen_fail` method. The writers of (future, if needed) children `_on_listen_fail` methods will have to call `super()` in their implementation. These are private methods and so we have full control of their code, I'd much rather have this caveat than two private methods that mostly do the same.

Reply:
I did it as catching `BaseException` is almost always the incorrect answer and generally not-desired by users either; I wanted to make it more difficult for users to accidentally mess something up as they expected `_on_listen_fail` to only catch `Exception`s, not things like `SystemExit` raised from `sys.exit(0)`. The other reason was catching `NotImpelementedError`, as I didn't think users should have to think about that at all, as they likely wouldn't expect this issue to arise.

I don't mind merging the two if you prefer though since this is a private method rather than part of the public API; i.e. I am ok with requiring users who subclass this to be better informed of the internals of this class than users of the public API. That wast just my rational for this setup.

With that said, would you prefer I merge them into a single `_one_listen_fail` function?

Reply:
I would still merge the two and expect users that subclass to be aware of the code.